desktop/core/src/desktop/middleware_test.py

```diff
@@ -156,3 +162,39 @@ def test_ensure_safe_redirect_middleware():
     settings.MIDDLEWARE.pop()
     for finish in done:
       finish()
+
+def test_spnego_middleware():
+  done = []
+  try:
+    # use SpnegoDjangoBackend to enable 'desktop.middleware.SpnegoMiddleware'
+    done.append(desktop.conf.AUTH.BACKEND.set_for_testing(['desktop.auth.backend.SpnegoDjangoBackend']))
+
+    c = Client()
+    with patch('kerberos.authGSSServerInit') as authGSSServerInit, \
+         patch('kerberos.authGSSServerStep') as authGSSServerStep, \
+         patch('kerberos.authGSSServerResponse') as authGSSServerResponse, \
+         patch('kerberos.authGSSServerClean') as authGSSServerClean, \
+         patch('kerberos.authGSSServerUserName') as authGSSServerUserName:
+      authGSSServerInit.return_value = 1, 'context'
+      authGSSServerStep.return_value = 1
+      authGSSServerResponse.return_value = 'gssstring'
+      authGSSServerClean.return_value = None
+      authGSSServerUserName.return_value = 'spnego_test'
+
+      header = {'HTTP_AUTHORIZATION': 'Negotiate test'}
+      response = c.get("/hue/editor/?type=impala", **header)
+      assert_equal(200, response.status_code)
```

Comment:
Either do 302 or better try `, follow=True` in the gets (to follow the redirects)

    self.test(*self.arg)
  File "/usr/share/hue/desktop/core/src/desktop/middleware_test.py", line 186, in test_spnego_middleware
    assert_equal(200, response.status_code)
AssertionError: 200 != 302

Reply:
Somehow, the AllowFirstUserDjangoBackend is getting called here. 
```
Traceback (most recent call last):
  File "/usr/share/hue/desktop/core/src/desktop/middleware.py", line 732, in process_request
    user = authenticate(username=username, request=request)
  File "/usr/share/hue/build/env/local/lib/python2.7/site-packages/Django-1.11.29-py2.7.egg/django/contrib/auth/__init__.py", line 70, in authenticate
    user = _authenticate_with_backend(backend, backend_path, request, credentials)
  File "/usr/share/hue/build/env/local/lib/python2.7/site-packages/Django-1.11.29-py2.7.egg/django/contrib/auth/__init__.py", line 116, in _authenticate_with_backend
    return backend.authenticate(*args, **credentials)
  File "/usr/share/hue/desktop/core/src/desktop/auth/backend.py", line 283, in authenticate
    password = kwargs['password']
KeyError: 'password'
```

And this seems not enough:
```
done.append(desktop.conf.AUTH.BACKEND.set_for_testing(['desktop.auth.backend.SpnegoDjangoBackend']))
```
I can reproduce the keyError if I add AllowFirstUserDjangoBackend in my local pseudo-distributed.ini file.

Is there way to remove AllowFirstUserDjangoBackend or force SpnegoDjangoBackend only?

Reply:
Updated with orig_backends = settings.AUTHENTICATION_BACKENDS